domain/modelconfig/service/service.go

```diff
@@ -273,19 +276,20 @@ func (s *Service) UpdateModelConfig(
 // secrets in dqlite
 type dummySecretsBackendProvider struct{}
 
-// dummySpaceProvider implements validators.SpaceProvider and always returns true.
-// TODO (tlm): These needs to be swapped out with an actual checker when we have
-// spaces in dqlite
-type dummySpaceProvider struct{}
-
 // HasSecretsBackend implements validators.SecretBackendProvider
 func (*dummySecretsBackendProvider) HasSecretsBackend(_ string) (bool, error) {
 	return true, nil
 }
 
-// HasSpace implements validators.SpaceProvider
-func (*dummySpaceProvider) HasSpace(_ string) (bool, error) {
-	return true, nil
+// spaceValidator implements validators.SpaceProvider.
+type spaceValidator struct {
+	st State
+}
+
+// HasSpace implements validators.SpaceProvider. It checks whether the
+// given space exists.
+func (v *spaceValidator) HasSpace(spaceName string) (bool, error) {
+	return v.st.CheckSpace(context.TODO(), spaceName)
```

Comment:
Make this method take a context. We can't have db queries not have a context, it can cause a potential lockup

Reply:
@SimonRichardson I added a todo comment, I had already started adding the context but it is a huge change (need to refactor all of environs config to add contexts) and wanted to do it in a separate PR. 
Is having this `context.TODO()` an issue for landing this while I finish putting up that PR?

Reply:
It needs to be done in this or the next, otherwise we'll forget about it.

Reply:
It is already ongoing, I started that work at the same time, but it's going to take longer than this PR. Also, as we have discussed previously, it's better to have smaller PRs than one huge PR that makes changes to different parts of juju

Reply:
@SimonRichardson https://github.com/juju/juju/pull/17330